opm/simulators/aquifers/AquiferNumerical.hpp

```diff
@@ -0,0 +1,229 @@
+        data.volume = this->cumulative_flux_;
+        data.type = Opm::data::AquiferType::Numerical;
+        return data;
+    }
+
+    void initialSolutionApplied()
+    {
+        this->init_pressure_ = this->calculateAquiferPressure();
+        this->pressure_ = this->init_pressure_;
+        this->flux_rate_ = 0.;
+        this->cumulative_flux_ = 0.;
+    }
+
+private:
+    const size_t id_;
+    const Simulator& ebos_simulator_;
+    double flux_rate_; // aquifer influx rate
+    double cumulative_flux_; // cumulative aquifer influx
+    const int* global_cell_; // mapping to global index
+    double init_pressure_;
+    double pressure_; // aquifer pressure
+
+    // TODO: maybe unordered_map can also do the work to save memory?
+    std::vector<int> cell_to_aquifer_cell_idx_;
+
+    double calculateAquiferPressure() const
+    {
+        double sum_pressure_watervolume = 0.;
+        double sum_watervolume = 0.;
+
+        ElementContext  elem_ctx(this->ebos_simulator_);
+        const auto& gridView = this->ebos_simulator_.gridView();
+        auto elemIt = gridView.template begin</*codim=*/0>();
+        const auto& elemEndIt = gridView.template end</*codim=*/0>();
+        for (; elemIt != elemEndIt; ++elemIt) {
+            const auto& elem = *elemIt;
+            if (elem.partitionType() != Dune::InteriorEntity) {
+                continue;
+            }
+            elem_ctx.updatePrimaryStencil(elem);
+
+            const size_t cell_index = elem_ctx.globalSpaceIndex(/*spaceIdx=*/0, /*timeIdx=*/0);
+            const int idx = this->cell_to_aquifer_cell_idx_[cell_index];
+            if (idx < 0) {
+                continue;
+            }
+
+            elem_ctx.updatePrimaryIntensiveQuantities(/*timeIdx=*/0);
+            const auto& iq0 = elem_ctx.intensiveQuantities(/*spaceIdx=*/0, /*timeIdx=*/0);
+            const auto& fs = iq0.fluidState();
+
+            // TODO: the porosity of the cells are still wrong for numerical aquifer cells
+            // Because the dofVolume still based on the grid information.
+            // The pore volume is correct. Extra efforts will be done to get sensible porosity value here later.
+            const double water_saturation = fs.saturation(waterPhaseIdx).value();
+            const double porosity = iq0.porosity().value();
+            const double volume = elem_ctx.dofTotalVolume(0, 0);
+            // TODO: not sure we should use water pressure here
+            const double water_pressure_reservoir = fs.pressure(waterPhaseIdx).value();
+            const double water_volume = volume * porosity * water_saturation;
+            sum_pressure_watervolume += water_volume * water_pressure_reservoir;
+            sum_watervolume += water_volume;
+        }
+
+        const auto& comm = this->ebos_simulator_.vanguard().grid().comm();
+        comm.sum(&sum_pressure_watervolume, 1);
+        comm.sum(&sum_watervolume, 1);
+        return sum_pressure_watervolume / sum_watervolume;
```

Comment:
Somehow, for parallel running, the data between aquifer 1 and aquifer 2 can be polluted here. Then different process obtain different aquifer pressure values through this function, at the end, the maximum was output into the summary file. Not understanding how this happens at all.

Reply:
It can be std::unordered_map gives different orders in different process, which makes the data between different aquifers mixed up together over different processes.  Will look for a solution tomorrow.